### PR TITLE
MAGE Dark Mode 1.0

### PR DESCRIPTION
**Pull request by ConConner**
## Adds full theming capabilities to MAGE.  
Due to the nature of WinForms this required the use of some custom controls and weird workarounds.
* Forms have special code that initializes the theme
* Minor adjustments on some forms
* Replacement of stock controls on each form with custom controls

Custom controls include:
* Flat TextBox
* Flat ComboBox
* Flat TabControl
* Flat NumericUpDown

A new Form exists, the ThemeEditor, which allows for the creation of custom themes.  
There are two standard themes: "Mage Old" and "Mage Dark", which cannot be deleted.  
Themes can be exported and imported as `MAGE Theme Files (.mtf)`.  

The custom code that is not integrated into Forms directly is relatively encapsulated in a new namespace called `Theming`

## Known Bugs
- List View coloumn headers are not themed
- Tree View structure is not themed
- Text Boxes have very intense flickering when hovering over them
  - This is a very tough one to fix
- Scroll bars are not themed
  - pretty much impossible to fix. Would require a custom scroll bar control
  - said control would need to be injected in every control that can display them, somehow

### CODE REVIEW

File: mage/FormShortcuts.cs [hunk 1]
```diff
@@ -14,6 +15,12 @@ public FormShortcuts(FormMain main)
         {
             InitializeComponent();
 
+            ThemeSwitcher.ChangeTheme(Controls, this);
+            ThemeSwitcher.InjectPaintOverrides(Controls);
+
+            ThemeSwitcher.ChangeTheme(this.Controls, this);
+            ThemeSwitcher.InjectPaintOverrides(this.Controls);
+
```
Comment (biosp4rk) [inline mage/FormShortcuts.cs:20]: Looks like this is unnecessary

File: mage/Theming/ThemeSwitcher.cs [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        private static ColorTheme projectTheme;
+        private static string projectThemeName;
+        public static string ProjectThemeName 
+        {
+            get => projectThemeName;
+            set
+            {
+                if (!Themes.ContainsKey(value))
+                {
+                    if (value == "Mage Old") Themes.Add("Mage Old", StandardTheme);
```
Comment (biosp4rk) [inline mage/Theming/ThemeSwitcher.cs]: You should add constants for theme names like "Mage Old" and "Mage Dark"

File: mage/Editors/FormPalette.cs [hunk 3]
```diff
@@ -294,21 +301,43 @@ private unsafe void DrawColorBar()
         private unsafe void DrawLine(BitmapData imgData, int color)
```
Comment (biosp4rk) [inline mage/Editors/FormPalette.cs:308]: Is it possible to call this function if the palette editor is open when the theme changes?

File: mage/FormShortcuts.cs (hunk 1 above)
Comment (ConConner) [reply mage/FormShortcuts.cs:20]: No this is necessary in every form to do the actual theme changing

Comment (biosp4rk) [reply mage/FormShortcuts.cs:20]: Look at the full context, it's a duplicate

Comment (ConConner) [reply mage/FormShortcuts.cs:20]: Oh you are right

**Pull request merged**